config/v1/types_descheduling.go

```diff
@@ -0,0 +1,45 @@
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Descheduler holds cluster-wide config information to run the Kubernetes Descheduler
+// and influence its placement decisions. The canonical name for this config is `cluster`.
+type Descheduler struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	// +required
+	Spec DeschedulerSpec `json:"spec"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status DeschedulerStatus `json:"status"`
+}
+
+type DeschedulerSpec struct {
+	// Policy contains the name of an upstream Descheduler policy configmap in the openshift-kube-descheduler-operator project
+	Policy ConfigMapNameReference `json:"policy"`
```

Comment:
Having a generic configmap means that your configuration API is effectively unbounded.  We have avoided unbounded configuration APIs because it means that versions skew and upgrades become difficult to reason about and it means that customers can configure things that aren't safe to support.

Why would we choose to do this instead of building a stable and safe configuration API that corresponds to what we want to support?

Reply:
The configuration API is defined upstream, and the configmap holds that yaml (of type `DeschedulerPolicy`, see https://github.com/kubernetes-sigs/descheduler/#policy-and-strategies).

With the upstream API, trying to define our own configuration API has become unnecessarily redundant and chaotic when we could just consume what is already available

Reply:
I did ponder this topic for a while, I was considering what David is suggesting here to hide the details, I've also considered the approach we took with audit with exposing only a set of pre-defined profiles, but I couldn't come up with a reasonable option. If it's possible to define a set of pre-defined descheduler profiles I'm open to that proposition, but for now I'm ok with us exposing that policy directly.

Reply:
Maintaining our own configuration API has introduced more bugs, as well as created duplicate work for us since every new feature in descheduler has also needed to be added to our operator API. Given that there's currently no features we exclude, it would be easier to parse the upstream API and add exceptions if we ever hit that case.

Reply:
> Maintaining our own configuration API has introduced more bugs, as well as created duplicate work for us since every new feature in descheduler has also needed to be added to our operator API. Given that there's currently no features we exclude, it would be easier to parse the upstream API and add exceptions if we ever hit that case.

I would posit that if you provide every configuration option to a customer, you have recreated the situation that 4.x was built to solve. Customers don't need to configure every knob and in fact often configure them in ways that constrain future upgradeability and cluster stability.

Rather than say, "currently there isn't anything I'm worried about", you should consider how configuration moves together and how to summarize that intent without exposing each individual unit.  The profiles that we use in other locations for TLS, audit log, etc are good examples.

If you support the entire configuration in an opaque configmap, I don't think you can feasibly change that API contract while maintaining backward compatibility.